plugins/resource_monitor_plugin/resource_monitor_plugin.cpp

```diff
@@ -0,0 +1,231 @@
+
+      auto interval = options.at("resource-monitor-interval-seconds").as<uint32_t>();
+      EOS_ASSERT(interval > 0 && interval < 60, chain::plugin_config_exception,
+         "\"resource-monitor-interval-seconds\" must be greater than 0 and less than 60 (1 minute)");
+      sleep_time_in_ms = interval * 1000;  // Convert to milliseconds
+      ilog("interval = ${interval}", ("interval", interval));
+
+      used_space_threshold_in_percentage = options.at("resource-monitor-space-threshold").as<uint32_t>();
+      EOS_ASSERT(used_space_threshold_in_percentage > 0 && used_space_threshold_in_percentage < 100, chain::plugin_config_exception,
+         "\"resource-monitor-space-threshold\" must be greater than 0 and less than 100");
+      ilog("threshold = ${threshold}", ("threshold", used_space_threshold_in_percentage));
+
+      // Add filesystems containing the directories to the monitored list
+      // and set up actual threshold
+      for ( auto& dir: dirs ) {
+         add_file_system( dir );
+      }
+   }
+
+   // Start main thread
+   void plugin_startup() {
+      ilog("Creating and starting monitor thread");
+      monitor_thread = std::thread( [this] {
+         fc::set_os_thread_name( "resmon" ); // console_appender uses 9 chars for thread name reporting. 
+         monitor_task();
+      } );
+   }
+
+   // System is shutting down.
+   void plugin_shutdown() {
+      ilog("shutdown...");
+
+      // Signal the monior thread to shut down
+      done = true;
+
+      // Wait for the thread to end
+      monitor_thread.join();
+
+      ilog("exit shutdown");
+   }
+
+private:
+   std::thread               monitor_thread;
+   std::atomic_bool          done {false};
+
+   static constexpr uint32_t def_sleep_time_in_secs = 2;
+   static constexpr uint32_t def_used_space_threshold_in_percentage = 90;
+
+   uint32_t                  sleep_time_in_ms {def_sleep_time_in_secs * 1000}; 
+   uint32_t                  used_space_threshold_in_percentage {def_used_space_threshold_in_percentage};
+
+   struct filesystem_info {
+      dev_t           st_dev; // device id of file system containing "file_path"
+      uintmax_t       min_available_bytes; // minimum number of available bytes the file system must maintain
+      bfs::path       path_name;
+   };
+
+   // Stores file systems to be monitored. Duplicate
+   // file systems are not stored.
+   std::vector<filesystem_info> filesystems;
+
+   void monitor_task() {
+      ilog("Entering monitor_task");
+
+      while ( true ) {
+         if ( is_any_filesystem_space_below_threshold() ) {
+            appbase::app().quit(); // This will gracefully stop Nodeos
+            break;
+         }
+         
+         std::this_thread::sleep_for( std::chrono::milliseconds( sleep_time_in_ms ));
+
+         if ( done ) {
+            break;
+         }
+      }
+   }
+
+   bool is_any_filesystem_space_below_threshold() {
```

Comment:
Can this be `const`?  Can we also separate this such that downstream effects of running out of space (eg, print a message and quit) are colocated and distinct from the method which checks the thresholds?

This would allow easier changes to how we handle this in the future and more precise testing efforts.

Reply:
Thanks Bart! I will print "quit" at the place I initiate quit, and remove "quit" print in is_any_filesystem_space_below_threshold. I will also think about ways to improve testability for future additions.